kornia/contrib/kmeans.py

```diff
@@ -0,0 +1,172 @@
+
+        Args:
+            X: 2D input tensor to be clustered
+            num_clusters: number of desired cluster centers
+
+        Returns:
+            2D Tensor with num_cluster rows
+        """
+        num_samples: int = len(X)
+        perm = torch.randperm(num_samples)
+        idx = perm[:num_clusters]
+        initial_state = X[idx]
+        return initial_state
+
+    def _pairwise_euclidean_distance(self, data1: Tensor, data2: Tensor) -> Tensor:
+        """Computes pairwise squared distance between 2 sets of vectors.
+
+        Args:
+            data1: 2D tensor of shape N, D
+            data2: 2D tensor of shape C, D
+
+        Returns:
+            2D tensor of shape N, C
+        """
+        # N*1*D
+        A = data1[:, None, ...]
+        # 1*C*D
+        B = data2[None, ...]
+        distance = (A - B) ** 2.0
+        # return N*C matrix for pairwise distance
+        distance = distance.sum(dim=-1)
+        return distance
+
+    def fit(self, X: Tensor) -> None:
+        """Iterative KMeans clustering till a threshold for shift in cluster centers or a maximum no of iterations
+        have reached.
+
+        Args:
+            X: 2D input tensor to be clustered
+        """
+        # X should have only 2 dimensions
+        KORNIA_CHECK_SHAPE(X, ["N", "D"])
+
+        if self.cluster_centers is None:
+            self.cluster_centers = self._initialise_cluster_centers(X, self.num_clusters)
+        else:
+            # X and cluster_centers should have same number of columns
+            KORNIA_CHECK(
+                X.shape[1] == self.cluster_centers.shape[1],
+                f"Dimensions at position 1 of X and cluster_centers do not match. \
+                {X.shape[1]} != {self.cluster_centers.shape[1]}",
+            )
+
+        # X = X.to(self.device)
+        current_centers = self.cluster_centers
+
+        previous_centers: Tensor | None = None
+        iteration: int = 0
+
+        while True:
+            # find distance between X and current_centers
+            distance: Tensor = self._pairwise_euclidean_distance(X, current_centers)
+
+            cluster_assignment = torch.argmin(distance, dim=1)
+
+            previous_centers = current_centers.clone()
+
+            for index in range(self.num_clusters):
+                selected = torch.nonzero(cluster_assignment == index).squeeze()
+                selected = torch.index_select(X, 0, selected)
+                # edge case when a certain cluster centre has no points assigned to it
+                # just choose a random point as it's update
+                if selected.shape[0] == 0:
+                    selected = X[torch.randint(len(X), (1,))]
+                current_centers[index] = selected.mean(dim=0)
+
+            # sum of distance of how much the newly computed clusters have moved from their previous positions
+            center_shift = torch.sum(torch.sqrt(torch.sum((current_centers - previous_centers) ** 2, dim=1)))
```

Comment:
probably here we can reuse the euclidean distance routine if we adjust the function to pass the reduction axis ? @johnnv1

Reply:
worth a try